doc/design/reader/low_level_reader.md

```diff
@@ -0,0 +1,71 @@
+# Paddle Low Level Data Reader
+
+This design doc will describe a way to implement C++ written data reader(feeder) to
+feed data to paddle when train or inference inorder to improve performance, and make
+best usage of CPU and GPU usage.
+
+This design can be used for both V2 API and new operator design.
+
+## Background
+
+We have sevaral issues talking about training with v2 API get bad training performance,
+like:
+- https://github.com/PaddlePaddle/Paddle/issues/3675 and
+- https://github.com/PaddlePaddle/Paddle/issues/4156 and
+- https://github.com/PaddlePaddle/Paddle/issues/3600.
+
+The current implementations for data reader can greatly simplify user defining how
+to parse and feed the data to neural network, but we need to copy data from python
+to C++ when feeding the data, and the data feeder will wait until the calculations
+of one mini-batch is done.
+
+In V1 implementations there are some considerations to improve data feeding
+performance, we can port these features to improve the reader.
+
+## Implementaion
+
+A low-level reader contains below components:
+
+1. Data parser:
+    A data parser will use `dlopen` to open user-defined parser plugin, and call
+    `parse` to convert a line of string to internal `Matrix` or `Tensor` object.
+1. User parser plugin:
+    User must write a C++ plugin to parse data to `Matrix` or `Tensor` object. The
+    plugin should implement one interface:
+    ```c++
+    // return Matrix* for v2 API call
+    Tensor* parse(const std::string& line);
+    ```
+1. Buffer:
+    A `DoubleBuffer` which is able to async load data when caculations are running.
```

Comment:
I think we can have a queue OP, and connect the reader OP to the queue OP. In this way the reader OP does not need to implement double buffer which duplicates functionality with the queue OP.

Reply:
Yep. This is a better design. I should have split the design for the layer-based reader and op-based reader.

Also, the queue op can be used for queueing messages between multi-device and multi-node training.